src/source/file/file_server.rs

```diff
@@ -56,15 +55,106 @@ impl Default for FileServerConfig {
+    // given loop. This cap grows each time we fail to read lines in an
+    // exponential fashion to some hard-coded cap.
+    loop {
+        let mut global_lines_read: usize = 0;
+        // glob poll
+        for entry in glob(pattern.to_str().expect("no ability to glob"))
+            .expect("Failed to read glob pattern")
+        {
+            if let Ok(path) = entry {
+                let entry = fp_map.entry(path.clone());
+                if let Ok(fw) = FileWatcher::new(&path) {
+                    entry.or_insert(fw);
+                };
+            }
+        }
+        // line polling
+        for (path, mut watcher) in fp_map.drain() {
+            let mut lines_read: usize = 0;
+            while let Ok(sz) = watcher.read_line(&mut buffer) {
+                if sz > 0 {
+                    lines_read += 1;
+                    lines.push(
+                        metric::LogLine::new(
+                            path.to_str().expect("not a valid path"),
+                            &buffer,
+                        ).overlay_tags_from_map(&config.tags),
+                    );
+                    buffer.clear();
+                } else {
+                    break;
+                }
+                if lines_read > config.max_lines_read {
+                    break;
+                }
+            }
+            report_full_telemetry(
+                "cernan.sources.file.lines_read",
+                lines_read as f64,
+                Some(vec![
+                    ("file_path", path.to_str().expect("not a valid path")),
+                ]),
+            );
+            // A FileWatcher is dead when the underlying file has
+            // disappeared. If the FileWatcher is dead we don't stick it in
+            // the fp_map_alt and deallocate it.
+            if !watcher.dead() {
+                fp_map_alt.insert(path, watcher);
+            }
+            global_lines_read = global_lines_read.saturating_add(lines_read);
+        }
+        for l in lines.drain(..) {
+            send(&mut chans, metric::Event::new_log(l));
+        }
+        // We've drained the live FileWatchers into fp_map_alt in the line
+        // polling loop. Now we swapped them back to fp_map so next time we
+        // loop through we'll read from the live FileWatchers.
+        mem::swap(&mut fp_map, &mut fp_map_alt);
+        // When no lines have been read we kick the backup_cap up by twice,
+        // limited by the hard-coded cap. Else, we set the backup_cap to its
+        // minimum on the assumption that next time through there will be
+        // more lines to read promptly.
+        if global_lines_read == 0 {
+            let lim = backoff_cap.saturating_mul(2);
+            if lim > 2_048 {
+                backoff_cap = 2_048;
+            } else {
+                backoff_cap = lim;
+            }
+        } else {
+            backoff_cap = 1;
+        }
+        let backoff = backoff_cap.saturating_sub(global_lines_read);
```

Comment:
What's the idea here? Seems to me this value will always either be `backoff_cap` or `0`:
* if global_lines_read > 0, then backoff_cap is 1. saturating_sub will produce 0.
* if global_lines_read = 0, then the subtraction does nothing.

Reply:
This is a block of code which was moved. 

@blt Can you please provide feedback to the original intent of this line?

Reply:
Huh, yeah, seems like some dodgy math. Good catch @dp.